File: ts/src/algorithms/post_order_dfs_binarytree.ts
import { BinaryTreeNode } from '../utils/binary_tree';
/**
 *  This code describes the recursive post order traversal
    of a binary tree. Before any given node in the tree is visited,
    all the nodes in the left subtree of the current node, and all
    the nodes in the right subtree of the current node are visited.
    
 * @param {object} node Root of Binary Tree
 * @param {number} node.val Value of current node 
 * @param {object} node.left Left subtree rooted to current node
 * @param {object} node.right Right subtree rooted to current node 
 * @param {number[]} output Array of values 
 * @returns {number[]} Output values in binary tree in post order 
 */
function postOrderDFSRecursive(
    node: BinaryTreeNode | null,
    output: number[],
): number[] | undefined {
    if (node == null) {
        return;
    }
    postOrderDFSRecursive(node.left, output);
    postOrderDFSRecursive(node.right, output);
    output.push(node.val);
    return output;
}

/**
 *  This code describes the iterative post order traversal
    of a binary tree. Before any given node in the tree is visited,
    all the nodes in the left subtree of the current node, and all
    the nodes in the right subtree of the current node are visited.

 * @param {object} node Root of Binary Tree
 * @param {number} node.val Value of current node 
 * @param {object} node.left Left subtree rooted to current node
 * @param {object} node.right Right subtree rooted to current node 
 * @returns {number[]} Output values in binary tree in post order 
 */
function postOrderDFSIterative(node: BinaryTreeNode): number[] {
    const output: number[] = [];
    const stack: BinaryTreeNode[] = [];
    let lastNode = null;
    // empty arrays, objects, sets are truthy statements in javascript
    // not in python so have to explicitly specify the length
    let mainNode: BinaryTreeNode | null = node;
    while (stack.length > 0 || node) {
        while (mainNode) {
            stack.push(mainNode);
            mainNode = node.left;
        }
        if (
            stack[stack.length - 1].right &&
            stack[stack.length - 1].right !== lastNode
        ) {
            mainNode = stack[stack.length - 1].right;
        } else {
            const currNode = stack.pop();
            lastNode = currNode;
            if (currNode) output.push(currNode.val);
        }
    }
    return output;
}

export { postOrderDFSIterative, postOrderDFSRecursive };

File: py/algorithms/post_order_dfs_binary_tree.py
""" This module contains code representing the post order depth
first search algorithm for a binary tree"""

from py.utils.binary_tree import binaryTreeNode
from typing import Literal, List


def post_order_dfs_recursive(node: binaryTreeNode,
                             output: Literal[[]]) -> List[int]:
    """ This code describes the recursive post order traversal
    of a binary tree. Before any given node in the tree is visited,
    all the nodes in the left subtree of the current node, and all
    the nodes in the right subtree of the current node are visited.

    Args:
        node:
            Object that represents the root of the binary tree to traverse
            in a postorder manner

        output:
            List of length zero that will store the output

    Returns:
        A list of integers containing the binary trees values in post order
    """
    if not node:
        return
    post_order_dfs_recursive(node.left, output)
    post_order_dfs_recursive(node.right, output)
    output.append(node.val)
    return output


def post_order_dfs_iterative(node: binaryTreeNode) -> List[int]:
    """ This code describes the iterative post order traversal
    of a binary tree. Before any given node in the tree is visited,
    all the nodes in the left subtree of the current node, and all
    the nodes in the right subtree of the current node are visited.

    Args:
        node:
            Object that represents the root of the binary tree to traverse in
            a postorder manner

    Returns:
        List of integers containing the binary trees values in post order
    """
    if not node:
        return
    stack = []
    output = []
    last_node = None
    while stack or node:
        while node:
            stack.append(node)
            node = node.left
        if stack[-1].right and stack[-1].right != last_node:
            node = stack[-1].right
        else:
            curr_node = stack.pop()
            output.append(curr_node.val)
            last_node = curr_node
    return output
